### ISSUE TITLE
Shall we base Foldable on `Iterator`?

### ISSUE THREAD

**Issue by julien-truffaut** [optics-dev/Monocle#1013]
Currently `Fold` is based on `foldMap` 

```scala
trait Fold[S, A] {
  def foldMap[M: Monoid](f: A => M)(s: S): M
}
```

The problem is that `foldMap` is not lazy, so when we use `foldMap` to implement `headOption` or `find`, we need to traverse the entire dataset.

```scala
val list = List(1, 2, 3, 4)
var counter = 0

val result = Fold.fromFoldable[List, Int].foldMap{x => counter +=1; Option(x)}(list)(Monoids.firstOption)

assertEquals(result, Some(1))
assertEquals(counter, 4)
```

`counter` is equal to 4, we traverse the entire list.

An alternative encoding could be based on `Iterator` which would allow a simple and lazy implementation for `headOption`, `find`, and so on.

```scala
trait Fold[S, A] {
  def iterator(from: S): Iterator[A]
}
```

**Comment by yilinwei** [comment]
I think this would be better - non-laziness for this type of operation is really unexpected.

**Comment by julien-truffaut** [comment]
@yilinwei Yeah I agree. Are you ok with exposing an `Iterator`? I think it is simple and does the job.

**Comment by yilinwei** [comment]
I took a quick look at the issues in cats, since I remember they had some issues with it.

I don't _think_ we'll run into this issue as long as `S => Iterator[A]` returns a fresh instance of an iterator.

[1] https://github.com/typelevel/cats/issues/1716
